include/envoy/http/async_client.h

```diff
@@ -38,7 +38,40 @@ class AsyncClient {
   };
 
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
+   */
+  class StreamCallbacks {
+  public:
+    virtual ~StreamCallbacks() {}
+
+    /**
+     * Called when the async HTTP stream receives headers.
+     * @param headers the headers received
+     * @param end_stream whether the response is header only
+     */
+    virtual void onHeaders(HeaderMapPtr&& headers, bool end_stream) PURE;
+
+    /**
+     * Called when the async HTTP stream receives data.
+     * @param data the data received
+     * @param end_stream whether the data is the last data frame
+     */
+    virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
I think the simple interface that @lizan wrote is the one to use. Here is why this all works this way: The way AsyncClient is implemented is that it basically mimics ActiveStream inside the HTTP connection manager, and instantiates a router filter to do the work. On one hand, this is kind of hack, on the other hand, it's an implementation detail so it doesn't really matter. So, in the end of the data, internally, the async client implements the StreamXXXCallback methods, which is the write interface, and then translates those "writes" into async client request/stream callbacks, which is this interface.